lib/common/format.h

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef FORTRAN_COMMON_FORMAT_H_
+#define FORTRAN_COMMON_FORMAT_H_
+
+#include "Fortran.h"
+#include <cstring>
+
+namespace Fortran::common {
+
+struct FormatError {
+  static constexpr int maxArgLength_{25};
+
+  const char *text_;  // error text; may have one %s argument
+  char arg_[maxArgLength_ + 1];  // optional %s argument value
+  int offset_;  // offset to error marker
+  int length_;  // length of error marker
+};
+
+template<typename CHAR = char> class FormatValidator {
+public:
+  explicit FormatValidator(const CHAR *format, size_t length, IoStmtKind stmt,
+      bool warnOnNonstandardUsage, FormatError *errors, int maxErrors)
+    : format_{format}, end_{format ? format + length : nullptr}, stmt_{stmt},
+      warnOnNonstandardUsage_{warnOnNonstandardUsage}, errors_{errors},
+      maxErrors_{maxErrors}, cursor_{format - 1} {}
```

Comment:
It seems like error reporting could be simplified by passing in a function to call to report errors.

Reply:
The administrative code for caller processing of errors is minimal.  As the code is now, I don't see much benefit to extracting single use functionality into a separate function.  If that changes as the code evolves, this could be reconsidered.

Reply:
> The administrative code for caller processing of errors is minimal.

The caller has to create a fixed-size array and create a validator where 3 of the 6 arguments are related to error processing. Then it has to loop through the errors and do something with them. That doesn't seem minimal to me.

> As the code is now, I don't see much benefit to extracting single use functionality into a separate function. If that changes as the code evolves, this could be reconsidered.

I don't understand what single use functionality you are referring to.

Reply:
The code has three parts - memory allocation, the Check call, and error reporting.  I believe memory allocation needs to be in the caller.  These two instances of that code in each of the Enter functions can probably be the same, but that is 3 (could be less) lines of code.   The Check calls and reporting code are different in the two Enter instances, and the kind=2/4 variants are just placeholders for currently unknown code.  When the eventual code is better understood, it may or may not make sense to put some of this code into a common function.

Reply:
I'm suggesting an interface to `FormatValidator` like this:
```
    common::FormatValidator<char> validator{formatCharBlock.begin(),
        formatCharBlock.size(), IoStmtKind::None,
        [&](int offset, int length, const char *text) {
          // report error
        }};
    validator.Check();
```

Reply:
There is now a FormatErrorReporter class derived from your suggestions (thanks!) that does at least some of this.